### PR TITLE
Updated ConversationalPipeline to work with encoder-decoder models

### PR DESCRIPTION
**Pull request by guillaume-be**
# What does this PR do?

This PR extends the capabilities of the existing ConversationalPipeline to work with Encoder Decoder models (such as BlenderBot). 

The pipeline has been modified as follows:
- history generated from concatenation of the inputs to the generated tokens for encoder-decoder (decoders use directly the generated tokens that contain the initial prompt)
- updated the cut-off position for generated tokens (1 for encoder-decoders, `input_length` for decoders)
- updated the clean-up script to remove all `pad_tokens` if `pad_token` != `eos_token`. Otherwise, remove pad_tokens from the second found if `pad_token` and `eos_token` are identical (previous behaviour of the pipeline). This is needed otherwise the models with a specific `eos_token` will keep an unnecessary `pad_token` that affects generation quality for subsequent rounds.

This has been tested with the BlenderBot 90M model (requires https://github.com/huggingface/transformers/pull/8205), producing the following output: https://gist.github.com/guillaume-be/380b98ec1ef91d0f6e3add5914dd92ce

## Who can review?

 Text Generation: @patrickvonplaten, @TevenLeScao
 Blenderbot, Bart, Marian, Pegasus: @sshleifer

### CODE REVIEW

**Review by sshleifer — APPROVED** [review]

File: src/transformers/pipelines.py [hunk 1]
```diff
@@ -2517,6 +2530,8 @@ def _clean_padding_history(self, generated_tensor) -> List[List[int]]:
             is_previous_pad = False
             for token in sequence:
                 if token == self.tokenizer.pad_token_id:
+                    if self.tokenizer.pad_token_id != self.tokenizer.eos_token_id:
```
Comment (patrickvonplaten) [inline src/transformers/pipelines.py:2533]: is this related to this issue: https://github.com/huggingface/transformers/issues/8032?

**Review by patrickvonplaten — COMMENTED** [review]
Looks great! Can we maybe add one test with the small `BlenderbotModel` to `/home/patrick/hugging_face/transformers/tests/test_pipelines_conversational.py` ?

File: src/transformers/pipelines.py (hunk 1 above)
Comment (guillaume-be) [reply src/transformers/pipelines.py:2489]: I believe it is not. The previous code from the conversational pipeline was assuming that `eos_token_id` is always equal to `pad_token_id`, and therefore only deleting `eos_token_id` starting from the 2nd consecutive occurence (these are the "padding" eos). Generally, `eos_token_id` is not `pad_token_id` and in that case the padding needs to be removed from the fist `pad_token` on. 

The other issue seems to be related to the generation process itself (creation of incorrect attention masks) - this specific statement is a post-processing step after the generation finishes.

**Comment by guillaume-be** [comment]
> Looks great! Can we maybe add one test with the small `BlenderbotModel` to `/home/patrick/hugging_face/transformers/tests/test_pipelines_conversational.py` ?

Thank you @patrickvonplaten , I added an integration test using BlenderBot 90M.

**Review by patrickvonplaten — APPROVED** [review]
Awesome work! Thanks a lot for adding the test. Good to merge IMO @LysandreJik

**Review by LysandreJik — APPROVED** [review]
LGTM! Will wait for the tests to be green and merge.

**Comment by LysandreJik** [comment]
Merging, unrelated failure.

**Pull request merged**